auditlog/proxy_decorator.go

```diff
@@ -0,0 +1,132 @@
+	"bytes"
+	"io"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/ory/oathkeeper/driver/configuration"
+	"github.com/ory/oathkeeper/proxy"
+	"github.com/ory/x/logrusx"
+	log "github.com/sirupsen/logrus"
+)
+
+// RoundTripper interface is implemented by the Proxy structure and it's decorators.
+type RoundTripper interface {
+	RoundTrip(r *http.Request) (*http.Response, error)
+	Director(r *http.Request)
+}
+
+// ProxyAuditLogDecorator is a wrapper for Proxy struct with audit logging abilities.
+type ProxyAuditLogDecorator struct {
+	proxy    RoundTripper
+	builders []EventBuilder
+	senders  []Sender
+	logger   *logrusx.Logger
+}
+
+func NewProxyAuditLogDecoratorFromFile(proxy *proxy.Proxy, config configuration.Provider,
+	logger *logrusx.Logger) (*ProxyAuditLogDecorator, error) {
+	bs, err := DeserializeEventBuildersFromFiles(config.AuditLogConfigPath())
+	if err != nil {
+		return nil, err
+	}
+
+	return NewProxyAuditLogDecoratorFromEventBuilders(proxy, config, logger, bs)
+}
+
+func NewProxyAuditLogDecoratorFromEventBuilders(proxy *proxy.Proxy, config configuration.Provider,
+	logger *logrusx.Logger, bs []EventBuilder) (*ProxyAuditLogDecorator, error) {
+	d := &ProxyAuditLogDecorator{
+		proxy:    proxy,
+		builders: bs,
+		senders:  make([]Sender, 0),
+		logger:   logger,
+	}
+
+	d.senders = append(d.senders, &StdoutSender{l: logger})
+	if config.AuditLogKafkaEnabled() {
+		d.senders = append(d.senders, &KafkaSender{})
+	}
+
+	return d, nil
+}
+
+// RoundTrip performs wrapped structure's RoundTrip and logs request's event.
+func (d *ProxyAuditLogDecorator) RoundTrip(req *http.Request) (*http.Response, error) {
+	// Copy request body.
+	var reqBodyCopy io.ReadCloser = nil
+	if req != nil {
+		req.Body, reqBodyCopy = copyBody(req.Body, d.logger)
+	}
+
+	// Send request.
+	resp, err := d.proxy.RoundTrip(req)
+
+	// Copy response body.
+	var respBodyCopy io.ReadCloser = nil
+	if resp != nil {
+		resp.Body, respBodyCopy = copyBody(resp.Body, d.logger)
+	}
+
+	// Log event.
+	go d.saveEvent(req, resp, reqBodyCopy, respBodyCopy, err)
+
+	return resp, err
+}
+
+func copyBody(rc io.Reader, logger *logrusx.Logger) (a, b io.ReadCloser) {
+	if rc != nil {
+		body, err := ioutil.ReadAll(rc)
+		if err != nil {
+			logger.Error("Error reading request body in RoundTrip")
+		}
+
+		a = ioutil.NopCloser(bytes.NewReader(body))
+
+		b = ioutil.NopCloser(bytes.NewReader(body))
+	}
+
+	return a, b
+}
+
+// Director performs wrapped structure's Director.
+func (d *ProxyAuditLogDecorator) Director(r *http.Request) {
+	d.proxy.Director(r)
+}
+
+func (d *ProxyAuditLogDecorator) saveEvent(reqImmutable *http.Request, respImmutable *http.Response,
+	reqBodyCopy, respBodyCopy io.ReadCloser, roundTripError error) {
+	if reqImmutable == nil {
+		d.logger.Error("Request struct is nil")
+
+		return
+	}
+
+	if respImmutable == nil {
+		d.logger.Error("Response struct is nil")
+
+		return
+	}
+
+	// Deep copy request & response.
+	req := reqImmutable.Clone(reqImmutable.Context())
```

Comment:
I haven't done any research on this case, but is it possible that the request can be changed or something... Is it thread-safe to just copy the request this way? The same for the response.

Reply:
I didn't look into it too. It seems that none of the middlewares should change the request and response context, but I'll move the copy creation to the caller side to eliminate any possible data races. Creating a copy inside the `RoundTrip` method is 100% safe but is not as efficient as here.

Reply:
Done. Please look if it is OK for you now.

Reply:
If you sure that it's not possible to modify the request/response then it's ok for me to copy in the child goroutine. Or if it's fine to get possibly corrupted request/response inside AuditLog. It was just a question :) 

But I totally agree that creating a copy inside the `RoundTrip` would be safer.

BTW `go test -race ./...` / `go run -race .` doesn't mind copying inside the child goroutine.

Reply:
It's OK for me now.